driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -838,7 +838,7 @@ void openAsyncOrGetAvailable(
         private PooledConnection acquirePermitOrGetAvailableOpenedConnection(final boolean tryGetAvailable, final Timeout timeout)
                 throws MongoTimeoutException {
             PooledConnection availableConnection = null;
-            tryLock(timeout);
+            lockInterruptibly(lock);
```

Comment:
Why is it ok to ignore the timeout when acquiring `lock`?

Reply:
To some extent this is answered in the commit message (which I modified several times, which resulted in re-creating the only commit in this PR again and again). But since this is still not obvious, let's think about a situation when locking may take time significant enough for us to want to make the locking itself timed.

The time to acquire this fair reentrant lock is `numberOfContendingThreads * averageTimeBeforeCallingWait`. `numberOfContendingThreads` is the number of threads which are ahead of the considered thread in the imaginary queue of threads that concurrently want to acquire the lock; `averageTimeBeforeCallingWait` is usually the average time to check the condition, but in our case it's more than that.

`numberOfContendingThreads`
Most of the time this lock is uncontended because threads get opened connections, and only after invalidation of the pool (or after creation of the pool) the lock has a chance to be contended.

`averageTimeBeforeCallingWait`
Has a potential to take some time only if the pool has many stale connections, which may happen if it has recently been invalidated and has large `maxSize`. While any individual invocation of `getPooledConnectionImmediately` has a potential to be slightly time-consuming (and, therefore, occasionally acquiring the lock may take comparable time), on average this method is of constant small time because the majority of invocations observe no stale connections.

If we also take into account that
1. all timeout implementations always overtime the specified timeout and that
2. we have a much more significant way to overtime: all asynchronous check outs and all (internal to the pool) asynchronous open requests are queued (open requests are also throttled) because they are executed by a single thread

worrying about the time to acquire the lock here does not seem important enough anymore. These thoughts appeared to me only recently, and not when I was implementing the functionality.

Reply:
Thanks for the more detailed explanation.  And really, the problem was that I didn't even look at the commit message.  Usually with single commit PRs the commit message is copied to the PR description automatically.  Perhaps that didn't happen because, as you said, you changed the message a few times.